server/CGameHandler.cpp

```diff
@@ -4338,24 +4338,32 @@ void CGameHandler::useChargedArtifactUsed(const ObjectInstanceID & heroObjectID,
 	if(vstd::contains(hero->getSpellsInSpellbook(), spellID))
 		return;
 
-	std::vector<std::pair<ArtifactPosition, ArtifactInstanceID>> chargedArts;
+	std::optional<std::tuple<ArtifactPosition, ArtifactInstanceID, uint16_t>> chargedArt;
 	for(const auto & [slot, slotInfo] : hero->artifactsWorn)
 	{
 		const auto * artInst = slotInfo.getArt();
 		const auto * artType = artInst->getType();
-		if(artType->getDischargeCondition() == DischargeArtifactCondition::SPELLCAST)
+		if(artInst->hasBonusOfType(BonusType::SPELL, spellID))
 		{
-			chargedArts.emplace_back(slot, artInst->getId());
-		}
-		else
-		{
-			if(const auto bonuses = artInst->getBonusesOfType(BonusType::SPELL, spellID); !bonuses->empty())
+			if(artType->getDischargeCondition() == DischargeArtifactCondition::SPELLCAST)
+			{
+				if(const auto spellCost = artType->getChargeCost(spellID))
+				{
+					chargedArt.emplace(slot, artInst->getId(), spellCost.value());
+					break;
+				}
+				else
+					return;
+			}
+			else
+			{
 				return;
+			}
 		}
 	}
 
-	assert(!chargedArts.empty());
-	DischargeArtifact msg(chargedArts.front().second, 1);
-	msg.artLoc.emplace(hero->id, chargedArts.front().first);
+	COMPLAIN_RET_IF(!chargedArt.has_value(), "Hero does not have charged artifacts with the required spell");
```

Comment:
Might be better to throw runtime error.

If this happens, this is not incorrect input, but rather incorrect logic somewhere in server code